### PR TITLE
Reflect and (de)serialize resources in scenes

### LINKED ISSUE #3576: Use scenes to serialize and deserialize resources
## Problem

Serializing and deserializing resources is a common and important task. Currently, this must be done manually.

However, we already have a tool for this serialization: `Scenes`. Each `Scene` stores a `World`, which (now) includes resources, but the rest of the machinery is missing.

## Proposed solution

Update the `Scene` machinery to handle resources as well. This should be reasonably straightforward (we can reflect resources, and we get access to the whole world when loading / saving / storing scenes), but fleshing out the rest of the API will be a fair bit of work.

The hardest part will be updating the `serde` `.ron` format: this presumes that only entities are included.

Our [scene example](https://github.com/bevyengine/bevy/blob/main/examples/scene/scene.rs)  should also be updated to show how to serialize / deserialize resources.

## Alternatives considered

We could make a distinct serialization strategy for resources. This makes sense to include as part of scenes because:

1. Common uses of scenes (saving / loading games, grabbing scenes from editors, networking) often include critical resources (current score, global illumination settings, difficulty settings) that should be logically batched together.
2. Refactoring between entities and resources is very common, and two seperate strategies would cause frustrating refactoring pain.
3. We're already storing a `World` everywhere for scenes.

## Context

Before #1144, the `World` only stored entities and components. This code hasn't been updated to work for resources too yet.

### PR DESCRIPTION
**Pull request by Davier**
# Objective

Fixes #3576.
Enables editors/inspectors to show and edit resources.

## Solution

- change the scene format to include resources:
```rust
(
  resources: [
    // List of resources here, same fomat as an entity's list of components
  ],
  entities: [
    // Previous scene format here
  ],
)
```
- enable reflection on a resource type (by adding `ReflectResource`)
```rust
#[derive(Reflect)]
#[reflect(Resource)]
struct MyResourse;
```
- add a vec of reflected resources to `DynamicScene`s and implement their (de)serialization
- adapt `Scene`s and `DynamicScene`s to actually import/export reflected resources from the world

## Remaining work

- [ ] write some tests
- [ ] modify the scene example to show resources too
- [x] gather feedback on a list of bevy's resources that should be reflected

## Notes

I ignored non send resources as I don't think they should be loaded from scenes, but it may make sense to reflect them so that they are accessible to editors/inspectors.

### CODE REVIEW

File: crates/bevy_scene/src/dynamic_scene.rs [hunk 1]
```diff
@@ -35,7 +36,28 @@ impl DynamicScene {
...
         let type_registry = type_registry.read();
 
-        for archetype in world.archetypes().iter() {
+        let mut archetypes = world.archetypes().iter();
+
+        // Empty archetype
+        let _ = archetypes.next().unwrap();
+
+        // Resources archetype
+        let resources_archetype = archetypes.next().unwrap();
```
Comment (alice-i-cecile) [inline crates/bevy_scene/src/dynamic_scene.rs:45]: Ugh, I don't like that this is the best way we have to get this...

Comment (alice-i-cecile) [reply crates/bevy_scene/src/dynamic_scene.rs:45]: Could we add a constant or dedicated method to fetch this?

File: crates/bevy_scene/src/dynamic_scene.rs [hunk 2]
```diff
@@ -35,7 +36,28 @@ impl DynamicScene {
...
+                .and_then(|registration| registration.data::<ReflectResource>());
+            if let Some(reflect_resource) = reflect_resource {
+                if let Some(resource) = reflect_resource.reflect_resource(world) {
+                    scene.resources.push(resource.clone_value());
+                }
+            }
+        }
+
+        // Other archetypes
+        for archetype in archetypes {
```
Comment (alice-i-cecile) [inline crates/bevy_scene/src/dynamic_scene.rs:60]: IMO this should have a comment calling out that we need to ensure that we iterate first. Iteration being stateful is a bit surprising at times, and swapping this code with the code above feels like it should be a harmless code re-organization change when it very much isn't.

File: crates/bevy_scene/src/dynamic_scene.rs [hunk 3]
```diff
@@ -78,9 +100,30 @@ impl DynamicScene {
...
+                .ok_or_else(|| SceneSpawnError::UnregisteredType {
+                    type_name: resource.type_name().to_string(),
+                })?;
+            let reflect_resource = registration.data::<ReflectResource>().ok_or_else(|| {
+                SceneSpawnError::UnregisteredResource {
+                    type_name: resource.type_name().to_string(),
+                }
+            })?;
+
+            // If the entity already has the given resource attached,
```
Comment (alice-i-cecile) [inline crates/bevy_scene/src/dynamic_scene.rs:119]:
```suggestion
            // If the world already contains an instance of the given resource
```

**Comment by alice-i-cecile** [comment]
This is looking very good. That was easier than I expected, and should be incredibly useful. If you can comfortably add a method / pattern to access resource / non-resource archetypes in a less fragile way that would be lovely, but I can live without it.

I'll work on compiling a list of resources to reflect now.

File: crates/bevy_scene/src/dynamic_scene.rs (hunk 1 above)
Comment (Davier) [reply crates/bevy_scene/src/dynamic_scene.rs:45]: There is one, I'll switch to using it separately from the loop. The first two iterations of the loop will do nothing since there are no entities in the empty and resource archetypes.

**Comment by alice-i-cecile** [comment]
`#[(derive(Resource))` would improve both finding and reflecting resources dramatically. Well for now, manual search it is.

# Reflect and Serialize

- add trait bounded methods to `State`: this is critical for a lot of game-saving use cases
- add trait bounded methods to `Events`
- `FrameTimeDiagnosticsState` and `LogDiagnosticsState`: might be useful?
- `LogSettings`: probably useful
- `Time`
- `SpriteSettings`
- `Msaa`
- `AmbientLight`
- `WgpuOptions`
- `ScheduleRunnerSettings`
- `WinitConfig`
- `ClearColor`
- `DefaultTaskPoolOptions`
- `ReportExecutionOrderAmbiguities`
- `WindowDescriptor`
- `AssetServerSettings`
- `FixedTimestepState`
- `Input<T>`: serialization could be handy for networking
- `RenderGraph`
- `PipelineCompiler`
- `RenderResourceBinding`
- `AssetRenderResourceBindings`
- `ActiveCameras`
- `WireframeConfig`
- `DefaultTextPipeline`
- `SceneSpawner`
- `FlexSurface`

# Just Reflect
- `Assets<T>`: I don't think we can store this state on disk reasonably
- `AssetServer`: ditto
- `EntityLabels`: Entity identifiers are opaque
- `IOTaskPool`, `AsyncComputeTaskPool`, `ComputeTaskPool`: can't be stored on disk'
- `TypeRegistryArc`: I think that storing this on disk is a bit too meta?
- `Audio<T>`: I think this isn't going to be useful without a loaded asset?
- `Windows` and `WinitWindows`: not meaningful to store

# Nothing
- `CiTestingConfig`: maybe?? I don't mind having reflect on this

**Comment by alice-i-cecile** [comment]
Overall, it looks like `Reflect` should probably just be a trait bound for resources: there's virtually always a valid use case for it.

Serialize is likely not appropriate for resources that are about working with assets or physical hardware (or when dealing with `Entity` identifiers), but probably fitting for everything else.

File: crates/bevy_scene/src/dynamic_scene.rs (hunk 1 above)
Comment (Davier) [reply crates/bevy_scene/src/dynamic_scene.rs]: Fixed in f43d667

File: crates/bevy_scene/src/dynamic_scene.rs (hunk 2 above)
Comment (Davier) [reply crates/bevy_scene/src/dynamic_scene.rs]: Fixed in f43d667

File: crates/bevy_ecs/src/world/mod.rs [hunk 4]
```diff
@@ -644,16 +644,21 @@ impl World {
         Some(unsafe { ptr.cast::<T>().read() })
     }
 
+    /// Returns `true` if a resource of a given [`TypeId`] exists. Otherwise returns `false`.
+    #[inline]
+    pub fn contains_resource_type_id(&self, type_id: TypeId) -> bool {
```
Comment (jakobhellermann) [inline crates/bevy_ecs/src/world/mod.rs:649]: I think this should take a `ComponentId` instead of a `TypeId` to support future dynamic resources.

I expect that most functions will have a rust-native version with generics that is easy to use, and a dynamic version taking ids which will be useful for scripting and editors, for which a `TypeId` may not work.

File: crates/bevy_scene/src/dynamic_scene.rs [hunk 5]
```diff
@@ -35,7 +36,27 @@ impl DynamicScene {
         let mut scene = DynamicScene::default();
         let type_registry = type_registry.read();
 
+        // Resources archetype
+        let resources_archetype = world.archetypes().resource();
+        for component_id in resources_archetype.components() {
+            let reflect_resource = world
+                .components()
+                .get_info(component_id)
+                .and_then(|info| type_registry.get(info.type_id().unwrap()))
```
Comment (jakobhellermann) [inline crates/bevy_scene/src/dynamic_scene.rs:45]: Maybe skip components without a type id instead of panicking

File: crates/bevy_scene/src/scene_spawner.rs [hunk 6]
```diff
@@ -154,6 +161,33 @@ impl SceneSpawner {
...
+            let resources_archetype = scene.world.archetypes().resource();
+            for component_id in resources_archetype.components() {
+                let component_info = scene
+                    .world
+                    .components()
+                    .get_info(component_id)
+                    .expect("component_ids in archetypes should have ComponentInfo");
+
+                // Resources that are not registered are ignored
+                if let Some(registration) = type_registry.get(component_info.type_id().unwrap()) {
```
Comment (jakobhellermann) [inline crates/bevy_scene/src/scene_spawner.rs:174]: same here, skip instead of unwrap

**Review by jakobhellermann — CHANGES_REQUESTED** [review]

File: crates/bevy_scene/src/dynamic_scene.rs (hunk 5 above)
Comment (Davier) [reply crates/bevy_scene/src/dynamic_scene.rs]: Done in 0fd72ae

File: crates/bevy_scene/src/scene_spawner.rs (hunk 6 above)
Comment (Davier) [reply crates/bevy_scene/src/scene_spawner.rs:174]: Done in 0fd72ae

File: crates/bevy_ecs/src/world/mod.rs (hunk 4 above)
Comment (Davier) [reply crates/bevy_ecs/src/world/mod.rs:649]: While I agree with the long term vision, all reflection in bevy currently works on `TypeId`s only. It will need a dedicated PR to move it to `ComponentId`s (which I might tackle at some point).
The corresponding functions for components have two versions, so I could also add a `fn contains_resource_id(&self, id: ComponentId)` in this PR even though I won't be using it.

Comment (jakobhellermann) [reply crates/bevy_ecs/src/world/mod.rs:649]: My reasoning for exposing the `ComponentId` variant is that you can use it to lookup a type id (by doing `components.get_resource_id` yourself) but you can't lookup a `ComponentId` using the `TypeId` variant.

**Comment by Davier** [comment]
> Overall, it looks like `Reflect` should probably just be a trait bound for resources: there's virtually always a valid use case for it.

It's possible to use `bevy_ecs` without `bevy_reflect`, so we can't do that. (Also, I've be told previously that reflection should be opt-in, not opt-out or mandatory.)

> Just Reflect

Reflection and serialization are closely related, I'm not sure we can have the first without the second with the current derive macro (as discussed on discord, but I'm repeating it here for others). I'll investigate.

Thank you @alice-i-cecile for the research! I'll start by reflecting the simple "config"-only resources, but I might leave the ones with actual content to their own PRs.

File: crates/bevy_ecs/src/world/mod.rs (hunk 4 above)
Comment (Davier) [reply crates/bevy_ecs/src/world/mod.rs]: Fair enough, added in 8503ab7

File: crates/bevy_ecs/src/reflect.rs [hunk 7]
```diff
@@ -122,6 +123,82 @@ impl<C: Component + Reflect + FromWorld> FromType<C> for ReflectComponent {
...
+        (self.reflect_resource)(world)
+    }
+
+    /// # Safety
+    /// This method does not prevent you from having two mutable pointers to the same data,
+    /// violating Rust's aliasing rules. To avoid this:
+    /// * Only call this method in an exclusive system to avoid sharing across threads (or use a
+    ///   scheduler that enforces safe memory access).
+    /// * Don't call this method more than once in the same scope for a given resource.
+    pub unsafe fn reflect_resource_mut<'a>(&self, world: &'a World) -> Option<ReflectMut<'a>> {
```
Comment (jakobhellermann) [inline crates/bevy_ecs/src/reflect.rs:159]: There should be a `reflect_resource_mut(&self, &mut World)` and a `reflect_resource_mut_unchecked(&self, &World)` IMO, otherwise you are forced to use the unsafe API even if you have exclusive world access.

(same for `ReflectComponent` actually but that is out of scope for this PR)

**Review by jakobhellermann — CHANGES_REQUESTED** [review]

**Comment by jakobhellermann** [comment]
For motivation, these changes (only `ReflectResource`, not the scene stuff) make this possible:
grafik


All that's needed is
```rust
#[derive(Reflect, Default)]
#[reflect(Resource)]
struct Msaa {
```

```rust
app.register_type::<Msaa>()
```

File: crates/bevy_ecs/src/reflect.rs (hunk 7 above)
Comment (Davier) [reply crates/bevy_ecs/src/reflect.rs]: Fixed in 39a28c2, I had missed that method from `ReflectComponent`.

**Comment by Davier** [comment]
I did a pass over every resource, and selected those that only contain configuration (i.e. no state), and that can be changed after startup: `Msaa`, `WireframeConfig`, `AmbientLight`, `DirectionalLightShadowMap`, `PointLightShadowMap`, `ClearColor`, and `GamepadSettings`.

I have not yet reflected `GamepadSettings` because I don't know if the "serialize" feature is still relevant or should be removed.

Reflecting `ClearColor` and `GamepadSettings` adds new dependencies on `bevy_reflect`, should they be optional dependencies?

I'm not sure these are useful to reflect: `CiTestingConfig` and `ReportExecutionOrderAmbiguities`.

These resources are only read during startup: `ScheduleRunnerSettings`, `AssetServerSettings`, `WgpuOptions`, `WinitConfig`, `DefaultTaskPoolOptions`, `WindowDescriptor`, and `LogSettings`. Should we reflect them even though modifications are without any effect? It could be useful but also misleading.

These resources contain state, so I think they deserve discussion on what part should or should not be reflected in their own PRs: `State<T>`, `Events<T>`, `LogDiagnosticsState`, `IoTaskPool`, `AsyncComputeTaskPool`, `ComputeTaskPool`, `ActiveCamera`, `DefaultTextPipeline`, `SceneSpawner`, `FlexSurface`, `Windows`, `WinitWindows`, `TypeRegistryArc`, `Audio<AudioSource>`, `Diagnostics`, `Input<T>`, (for `KeyCode`, `MouseButton`, `GamepadButton`), `Touches`, `Axis<T>`, for `GamepadAxis`, `GamepadButton`), `Gamepads`, `VisiblePointLights`, `FrameTimeDiagnosticsState`, `Time`, `FixedTimesteps`,

**Comment by jakobhellermann** [comment]
> Reflecting ClearColor and GamepadSettings adds new dependencies on bevy_reflect, should they be optional dependencies?

IMO they can just depend on it unconditionally, just like `bevy_asset`, `bevy_gltf` etc. do. You'll have it in your dependency tree anyways.

**Comment by jakobhellermann** [comment]
Currently this PR will put every reflectable resource into the scene, right? I don't think that this is a very good policy, as there are a lot of resources like `Input<KeyCode>` that you don't want in your serialized scene.

What do you think about splitting this PR into one for `ReflectResource` and another one for allowing (de-)serialization of resources in scenes?
There former is fully implemented and as far as I can tell ready to merge, while the latter needs more thought put into it.

**Comment by Davier** [comment]
> Currently this PR will put every reflectable resource into the scene, right? I don't think that this is a very good policy, as there are a lot of resources like `Input<KeyCode>` that you don't want in your serialized scene.

You are talking about using `DynamicScene::from_world(world: &World, type_registry: &TypeRegistryArc)` right? Indeed, that will copy all components and resources that are registered in the given `type_registry` (i.e. not the world's `type_registry`). I think it would be surprising if it didn't. 
I do agree that you usually don't want to save all the reflected resources, but I think you also don't want to save all the reflected components. I expect people to make a custom `type_registry`, and only register the components and resources they want to be included in the scene (so the default is not all of them, but none of them).
Do you think it would be enough to document it properly, and make an example that explicitly shows how to use a custom `type_registry`?

For your example of `Input<KeyCode>`, I think there are cases when it would makes sense to serialize it. For instance, if you automatically save a game state, you might want to save what the user was pressing to keep things consistent.

> What do you think about splitting this PR into one for `ReflectResource` and another one for allowing (de-)serialization of resources in scenes? There former is fully implemented and as far as I can tell ready to merge, while the latter needs more thought put into it.

Both are implemented, I just need some time to rebase, and write some examples and tests. I can split the PR in two if we don't have a consensus on the second part :slightly_smiling_face: (I'll probably be to busy to do it until the end of march though)

**Comment by alice-i-cecile** [comment]
>  I think it would be surprising if it didn't.

Agreed.

> Do you think it would be enough to document it properly, and make an example that explicitly shows how to use a custom type_registry?

IMO yes.

> I can split the PR in two if we don't have a consensus on the second part

I think this is a good idea regardless; the functionality is related but distinct, and optimizing for merge-ability is a good call.

**Comment by alice-i-cecile** [comment]
@Davier if you have the chance, can you split out the reflection changes from the "add resources to scenes"? The former is uncontroversial, but the latter needs more consideration.

**Comment by alice-i-cecile** [comment]
Very cool, but I think there's a lot of value in having a fully safe API for a relatively common task.

**Comment by alice-i-cecile** [comment]
Closing in favor of #5175 as it is much less controversial. We can tackle "should resources be stored in scenes" when we tackle more comprehensive design reviews of how scenes and prefabs work.
